src/components/radio-group/radio-group.ts

```diff
@@ -39,6 +59,48 @@ export default class SlRadioGroup extends LitElement {
     });
   }
 
+  getAllRadios(): SlRadio[] {
+    return [...this.querySelectorAll('sl-radio')];
+  }
+
+  handleKeyDown(event: KeyboardEvent) {
+    if (['ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight'].includes(event.key)) {
+      const radios = this.getAllRadios().filter(radio => !radio.disabled);
+      const currentIndex = radios.findIndex(el => el.checked);
+
+      const incr = ['ArrowUp', 'ArrowLeft'].includes(event.key) ? -1 : 1;
+      let index = currentIndex + incr;
+      if (index < 0) index = radios.length - 1;
+      if (index > radios.length - 1) index = 0;
+
+      this.getAllRadios().map(radio => (radio.checked = false));
+      radios[index].focus();
+      radios[index].checked = true;
+
+      event.preventDefault();
+    }
+  }
+
+  reportValidity() {
+    const radios = [...(this.defaultSlot.assignedElements({ flatten: true }) as SlRadio[])];
+    let isChecked = true;
+
+    if (this.required && radios.length > 0) {
+      isChecked = radios.some(el => el.checked);
+
+      if (!isChecked) {
+        // This is hacky...
+        radios[0].required = true;
+
+        setTimeout(() => {
+          radios[0].reportValidity();
```

Comment:
The result of this call to `reportValidity()` should be the return value in case there's an inconsistency.

I'm assuming the `setTimeout` was added to wait for the next render so the radio will be ticked, right? All other `reportValidity()` methods in the library are synchronous, so making this one async feels a bit weird. This might be a good use case for calling [`update()`](https://lit.dev/docs/components/lifecycle/#update).

Does this work in lieu of the timeout?

```ts
radios[0].required = true;

const changedProps = new Map<string, any>();
changedProps.set('required', false);

radios[0].update(changedProps);
```

If not, we may have to live with making `reportValidity` an async method.

Reply:
The `setTimeout` was added because I ran into an issue with timing. 

I don't know when the `sl-radio` elements are added to the default slot. So I can't reliably force a required flag on the first radio element during initial render.

I just want the browser-level validation message to display on the first `sl-radio` element in the group, and the only way to trigger it is to set the underlying `InputElement` to `required`, then run the `reportValidation`. That requires that I manipulate the attribute on `sl-radio`, which means I need to wait for the event queue to flush before triggering the validation.

There are a couple of things I have thought about to get around this, but haven't decided:

1. Expose the underlying input element, modify it directly (very dirty IMHO).
2. Use a custom getter/setter for the `required` attribute, this would let me directly manipulate the input element before changes have trigger on the `sl-radio` prop change (which should cause the correct validity message to trigger right after I set the prop on `sl-radio`).
3. I could never expose ANY of the underlying input element (itself or attributes) and just provide a `triggerRequiredWarning` method of some sort. I didn't like this _at first_, but if you think about a use case where the first input might be off-screen (edge case IMHO)... `rl-radio-group` could handle that in the future and trigger the validation on a visible input. 
4. Something tired me isn't thinking about.

Reply:
@Trendy What about making the sl-radio responsible for checking its parent's required state? If a new/first sl-radio is added, it will itself acquire the correct state. If the sl-radio-group receives a (new) required state it can set it on the first sl-radio if present and otherwise assume the first child will pick it up itself.

Reply:
> Expose the underlying input element, modify it directly (very dirty IMHO).

In this case, it's cleaner than forcing an update I guess.

```js
radio.shadowRoot.querySelector('input[type="radio"]').required = true;
```

It would keep things synchronous, at least. I'm not opposed to it since the components are required to be used together.